stateproof/builder.go

```diff
@@ -201,7 +355,19 @@ func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.Forw
 				sfa.Round, proto.StateProofInterval)
 		}
 
-		builderForRound, err = spw.makeBuilderForRound(sfa.Round)
+		if sfa.Round > latest {
+			// avoiding an inspection in DB in case we haven't reached the round.
+			// Avoiding disconnecting the peer, since it might've been sent to this node while it recovers.
+			return network.Ignore, fmt.Errorf("handleSig: latest round is smaller than given round %d", sfa.Round)
+		}
+
+		// We want to save the signature in the DB if we know we generated it. However, if the signature's source is
+		// external, we only want to process it if we know for sure it meets our broadcast policy.
+		if sender != nil && !spw.meetsBroadcastPolicy(sfa, latestHdr.Round, &proto, stateProofNextRound) {
+			return network.Ignore, nil
+		}
+
+		builderForRound, err = spw.loadOrCreateBuilderWithSignatures(sfa.Round)
```

Comment:
Calling this here may eventually call `persistBuilder`. Could it happen that at the same time, `OnPrepareVoterCommit` also calls `persistBuilder` for the same round? In that case, one of them will error.